_posts/2024-06-28-overriding-configuration-from-test-code.adoc

```diff
@@ -0,0 +1,626 @@
+import io.restassured.RestAssured;
+import java.util.Map;
+import org.hamcrest.CoreMatchers;
+import org.junit.jupiter.api.Test;
+
+@QuarkusTest
+@TestProfile(FeaturesResourceTest.class)
+public class FeaturesResourceTest implements QuarkusTestProfile { // <1>
+
+    @Override
+    public Map<String, String> getConfigOverrides() { // <2>
+        return Map.of(
+            "features.awesome-feature-enabled", "true", // <3>
+            "amazing-feature-enabled", "true"
+        );
+    }
+
+    @Test
+    void test() {
+
+        RestAssured.given()
+            .when().get("/features/awesome")
+            .then().body(CoreMatchers.is("true"));
+
+        RestAssured.given()
+            .when().get("/features/amazing")
+            .then().body(CoreMatchers.is("true"));
+    }
+}
+----
+<1> The test class itself can implement `QuarkusTestProfile` if the profile isn't shared across multiple test classes.
+This can make the maintenance and reviews of the test code easier.
+If multiple test classes depend on the same profile, then that profile will likely need to be declared in a dedicated class.
+<2> This method comes from `QuarkusTestProfile` and makes it possible to override the configuration from the test code.
+<3> The config key generated from the `FeaturesConfig` interface is prefixed with `features.` while the config key that comes from the `@ConfigProperty` injection has no prefix.
+
+== Approach #2: mocking the config with Mockito
+
+Now, here's my favorite approach when native testing is not required.
+
+First, let's see how that works with a https://quarkus.io/guides/config-mappings[config mapping^]:
+
+[source, java, title=The code to be tested]
+----
+import io.smallrye.config.ConfigMapping;
+import io.smallrye.config.WithDefault;
+import jakarta.inject.Inject;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+@Path("/features")
+public class FeaturesResource {
+
+    @Inject
+    FeaturesConfig featuresConfig;
+
+    @GET
+    @Path("/awesome")
+    public boolean isAwesomeFeatureEnabled() {
+        return featuresConfig.awesomeFeatureEnabled();
+    }
+}
+
+@ConfigMapping(prefix = "features")
+interface FeaturesConfig { // <1>
+
+    @WithDefault("false")
+    boolean awesomeFeatureEnabled();
+}
+----
+<1> In a real project, this interface would likely be `public` and declared in a separate file.
+
+[source, java, title=The test code]
+----
+import io.quarkus.test.InjectMock;
+import io.quarkus.test.Mock;
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
+import io.smallrye.config.SmallRyeConfig;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.inject.Produces;
+import jakarta.inject.Inject;
+import org.eclipse.microprofile.config.Config;
+import org.hamcrest.CoreMatchers;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+@QuarkusTest
+public class FeaturesResourceTest {
+
+    @Inject
+    Config config; // <1>
```

Comment:
You can inject `SmallRyeConfig` here directly, avoid the `unwrap` and the warning callout